fix: indentation issues

mbhall88 opened the pull request:
This PR addresses the indentation issues identified in #124. Specifically, it fixes https://github.com/snakemake/snakefmt/issues/124#issuecomment-982690694 and https://github.com/snakemake/snakefmt/issues/124#issuecomment-986845398

I have also adjusted the line spacing in some existing tests which I realised were incorrect. In particular, [black only inserts a single line around indented function/class definitions](https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#empty-lines). By extention, snakemake directives should not be given two empty lines when inside `if` blocks. As an example, black turns this

```python
if True:

    class A:
        pass


    x = "foo"


    class B:
        pass


class C:
    pass


x = "foo"


class D:
    pass
```

into 

```python
if True:

    class A:
        pass

    x = "foo"

    class B:
        pass


class C:
    pass


x = "foo"


class D:
    pass
```

I have not been able to fix all of these tests though due to extra new lines being added to code from black. These extra newlines get added because the context in which we pass the code to black is that it *isn't* indented, so for classes/function (i.e. snakemake directives), black thinks they are module-level, and thus gives them two empty lines.

mbhall88 commented:
Alright @bricoletc this is ready for a review. It isn't "pretty", but the patches do the job. The failing CI stuff is just shitty python/pip/mac arch-reaching stuff. Not actually tests failing.

bricoletc commented on snakefmt/formatter.py:
```diff
@@ -153,8 +153,12 @@ def run_black_format_str(
                 tmpstring += line
             string = textwrap.dedent(tmpstring)
 
+        needs_artifical_nest = inside_nested_statement and not string.startswith("if")
+        if needs_artifical_nest:
+            string = f"if x:\n{textwrap.indent(string, TAB)}"
+
```
I see, here you make sure `black` will indent python code with one newline only if we're in a nested context. Nice.

bricoletc replied:
I looked into the `black` `Mode` class that we tap into to modify line length, in case there was a way to tell black we're in a nested context - no such. I think your fix is good enough in that respect

mbhall88 replied:
Yeah, it feels so hacky, but as you say, there's no other way of passing that indentation context to black (which makes sense I guess).

bricoletc reviewed (approved):
Great work @mbhall88 !

I will push minor edits to this PR and request them for review by you.

bricoletc commented:
Hmm. I thought that pushing to this PR would allow a kind of overlay-diff, of my new commits to tip of your PR. 
Here's a summary of what I added:

* Bugfix: Hard-coded tab length - you used the number '4' as the length of a TAB, I replaced by len(TAB)
* Bugfix: indent level passed to black str formatting. Fixes my comment above
* Refactor: 'after_if' testing happens twice, refactored it as `+after_if_keywords = ("elif", "else")` in formatter.py

That;s for the PR

I also use `neovim` now and so I added `pynvim` to `pyproject.toml` in dev dependencies. If you'd rather add that at a later point, because unrelated to this PR, feel free to force-delete that commit.
Also it seems we can manage dependencies as [groups now in poetry](https://python-poetry.org/docs/managing-dependencies/#dependency-groups), maybe we should have a 'vim' dependency group or something. But pynvim is a small package, i dont think its a pain to install it by default.

bricoletc commented:
@mbhall88 I have made a fix to #159 in a new branch https://github.com/snakemake/snakefmt/tree/indent_overhaul
That branch is based off of tip of this PR. I will rebase it if and when this PR changes.

That branch passes the test you introduced replicating the #159 bug, but now 5 pre-existing tests fail.
More work required to address. I've simplified the pycode/snakecode parsing logic, which is a good start! - was too hacky before.

mbhall88 commented on snakefmt/formatter.py:
```diff
@@ -196,6 +205,10 @@ def align_strings(self, string: str, target_indent: int) -> str:
         indented = ""
         for match in re.finditer(full_string_matcher, string):
             indented += textwrap.indent(string[pos : match.start(1)], used_indent)
+            lagging_spaces = len(indented) - len(indented.rstrip(" "))
```
The issue was that the first line `datasources=(` has a trailing tab, which is fine, this causes the line starting with `"-s"` to be indented correctly, but the other lines **don't** have a trailing tab, and so the indentation for the other lines was not large enough (they're in one tab from the `"-s"` line). I tried a lot of things to try and fix this, but this was the only solution that I could get working in the end.  
Very happy if you have a more robust solution?

bricoletc replied:
:+1:  I see; i'll maybe have a look if we can do better, but not top-priority I think

mbhall88 commented:
Great changes/fixes.

> I also use neovim now and so I added pynvim to pyproject.toml in dev dependencies. If you'd rather add that at a later point, because unrelated to this PR, feel free to force-delete that commit.
Also it seems we can manage dependencies as [groups now in poetry](https://python-poetry.org/docs/managing-dependencies/#dependency-groups), maybe we should have a 'vim' dependency group or something. But pynvim is a small package, i dont think its a pain to install it by default.

No problems. Yeah, I noticed the change to the groups thing. I think maybe just best to do that with a separate push straight to master after merging this (and other?) PRs?

> That branch passes the test you introduced replicating the https://github.com/snakemake/snakefmt/issues/159 bug, but now 5 pre-existing tests fail.
More work required to address. I've simplified the pycode/snakecode parsing logic, which is a good start! - was too hacky before.

Hahaha these bug fixes are seriously like whack-a-mole with the tests. You get one passing and 6 others fail hahaha.

mbhall88 commented:
> @mbhall88 I have made a fix to https://github.com/snakemake/snakefmt/issues/159 in a new branch https://github.com/snakemake/snakefmt/tree/indent_overhaul
That branch is based off of tip of this PR. I will rebase it if and when this PR changes.

@bricoletc sorry it wouldn't allow me to do a rebase merge, so I had to do a merge commit. Shouldn't be the end of the world for your branch though. Worst case we just do another merge.

Feel free to open a draft PR for that branch if you want? Should I close #166 in favour of your new branch?

bricoletc commented:
No worries, I'll try rebasing my branch on top of master
And yes #166 should now be closed, and I will create a draft PR when I've done a few more things on that branch!
